### PR TITLE
add unarchiver example

### LINKED ISSUE #311: Easier unarchiving support with v4?
Hello, thanks for the awesome library. The new version is fantastic since I can use it in conjunction with other network libraries to stream HTTP response right into unarchiving process.

I'm using archiver on Google Cloud Storage caching library for GitHub Actions, and while archiving files was [super easy](https://github.com/shakrmedia/gcs-cacher/blob/minku/include-symlink/cacher/cacher.go#L138), unarchiving was [more involved](https://github.com/shakrmedia/gcs-cacher/blob/minku/include-symlink/cacher/cacher.go#L252) with some manual file type handling.

Is there a plan to have a shorthand to make unarchival process a bit easier?

## What would you like to have changed?

Quick way to unarchive file on v4, similar to how v3 works in some way

## Why is this feature a useful, necessary, and/or important addition to this project?

Currently, unarchiving requires a user to manually handle each different Tar type flags which involves a lot of boilerplate. 

## What alternatives are there, or what are you doing in the meantime to work around the lack of this feature?

I took a look at v3 code and [implemented unarchiving in-line](https://github.com/shakrmedia/gcs-cacher/blob/minku/include-symlink/cacher/cacher.go#L252), however it's not ideal at least from my perspective.

### PR DESCRIPTION
**Pull request by jm33-m0**
Hi Matt, thanks for the great work!

I noticed there's no `Unarchive` function in v4 yet, and the documentation is incomplete and hard to follow, so I wrote a small program that invokes your extractor API to automatically identify the format and unarchive tarballs.

Please check the code and point out where I need to improve. It can serve as an example for users who seek a quick solution from this project.

### CODE REVIEW

**Comment by mholt** [comment]
Thanks for this; I like the idea of posting an example instead of prescribing one way to do it in the code itself.

I have a few suggestions on how to tweak this but I will have to come back and write those up later.

I suppose this might resolve #311.

**Comment by mholt** [comment]
Does this example take into account #391?

**Comment by jm33-m0** [comment]
> Does this example take into account #391?

In fact, my code doesn't retain the file/dir permissions. So I guess it will "solve" #391.

**Comment by mholt** [comment]
Gotcha. This is a good start, but I think file mode bits (permissions, etc) are important, so we might need to include those in a proper example.

**Comment by jm33-m0** [comment]
I have updated the code with necessary changes to address the directory permission issue, and I also added code to handle symlinks.

**Comment by mholt** [comment]
Thank you! However, I don't think we should be providing an actual program, just an example in the docs.

I'm actually about the archive this repository, but we can continue this discussion over in [mholt/archives](https://github.com/mholt/archives) if you would like.
